grader_systems/src/stage/valgrind.rs

```diff
@@ -0,0 +1,1123 @@
+        }
+
+        if let Some(v) = self.config.free_fill {
+            cmd.add_arg(format!("--free-fill=0x{:02X}", v));
+        }
+
+        if let Some(v) = &self.config.suppressions {
+            for supp in v {
+                cmd.add_arg(format!("--suppressions={}", supp));
+            }
+        }
+
+        cmd.add_arg("--");
+        cmd.add_arg(&self.executable);
+        cmd.add_args(&self.args);
+        if let Some(v) = &self.stdin {
+            cmd.set_stdin(StdinPipe::Path(v.into()))
+        }
+
+        cmd.set_timeout(self.timeout.unwrap_or(DEFAULT_TIMEOUT));
+        cmd.set_cwd(ws);
+
+        cmd
+    }
+
+    fn read_logfile(&self, ws: &Path) -> Result<String> {
+        let path = ws.join(&self.config.log_file);
+        if !path.exists() {
+            return Err(anyhow!(
+                "Could not find logfile at {:?}. Did Valgrind actually execute?",
+                path.to_str()
+            ));
+        }
+        let contents = read_file(&path)?;
+        if contents.trim().is_empty() {
+            return Err(anyhow!(
+                "Found empty valgrind log at {:?}. Something went wrong.",
+                path.to_str()
+            ));
+        }
+
+        // replace all absolute paths with basename
+        let re = Regex::new(r"(\W|^)(?:\/[^\/\s]+)+\/([^\/\s]+)\b")?;
+        let repl = re.replace_all(&contents, "$1$2");
+
+        Ok(repl.to_string())
+    }
+}
+
+#[async_trait]
+impl<E: ProcessExecutor> Executor for Valgrind<E> {
+    type Output = StageResult;
+
+    async fn run(&self, ws: &Path) -> Result<Self::Output> {
+        let mut sect = Section::new("Valgrind");
+        let mut results_sect = Section::new("Output:");
+        let mut run_updates = StatusUpdates::default();
+        debug!("running program");
+
+        let executable = ws.join(&self.executable);
+        if !executable.exists() {
+            return Err(anyhow!(
+                "Could not find student executable at {:?}",
+                executable.display()
+            ));
+        }
+
+        if let Some(v) = &self.config.suppressions {
+            for supp in v {
+                if !ws.join(&supp).exists() {
+                    return Err(anyhow!("Could not find valgrind suppressions at {}", supp));
+                }
+            }
+        }
+
+        let cmd = self.gen_cmd(ws);
+        sect.add_content(("Run Command", format!("{}", cmd).code()));
+
+        let res = cmd.run_with(&self.executor).await?;
+
+        let log = self.read_logfile(ws)?;
+
+        if !res.status.completed() {
+            match res.status {
+                ExitStatus::Timeout(to) => {
+                    run_updates.add_update(
+                        Update::new_fail("running valgrind", PointQuantity::FullPoints).notes(
+                            format!(
+                                "Your submission timed out after {} seconds :(",
+                                to.as_secs()
```

Comment:
np, you can use `{:?}` to use debug formatting and Duration will automaticaly format the time to something human readable. For example, Duration::from_secs(2) will be printed as 2s

Reply:
Ah thanks! Good to know. Will probably keep the current version since "seconds" is probably appropriate

Reply:
yeah, sounds good.